src/quick-lint-js/fe/parse.cpp

```diff
@@ -314,6 +314,51 @@ void parser::error_on_sketchy_condition(expression* ast) {
   }
 }
 
+void parser::error_on_pointless_string_compare(
+    expression::binary_operator* ast) {
+  auto is_comparison_operator = [](string8_view s) {
+    return s == u8"=="sv || s == u8"==="sv || s == u8"!="sv || s == u8"!=="sv;
+  };
+
+  for (int i = 0; i < ast->child_count() - 1; i++) {
+    expression* lhs = ast->child(i);
+    expression* rhs = ast->child(i + 1);
+
+    if ((lhs->kind() == expression_kind::call &&
+         rhs->kind() == expression_kind::literal) ||
+        (lhs->kind() == expression_kind::literal &&
+         rhs->kind() == expression_kind::call)) {
+      source_code_span op_span = ast->operator_spans_[i];
+      if (!is_comparison_operator(op_span.string_view())) {
+        continue;
+      }
+
+      // make sure the call is on the "left" and the literal on the "right"
+      if (lhs->kind() == expression_kind::literal) {
+        expression* tmp = lhs;
+        lhs = rhs;
+        rhs = tmp;
+      }
+
+      string8_view call =
+          lhs->child_0()->variable_identifier().span().string_view();
```

Comment:
Must fix: This is wrong. You're assuming that the call expression's callee is a dot expression or a variable reference. The following is legal code, but this call to `variable_identifier` will crash:

```javascript
// binary(call(index(dot(var o, m), literal)), literal)
o.m[0]() == 42
```

You should check that `lhs`'s callee (`child_0`) is a `dot` expression before calling `variable_identifier`.